File: avaldused/client-java/ee/cyber/eid/EidClient.java
package ee.cyber.eid;

import java.security.cert.X509Certificate;
import java.util.List;
import java.util.Scanner;

import javax.smartcardio.CardException;

import ee.cyber.eid.card.CardUtil;
import ee.cyber.eid.card.EstEID;
import ee.cyber.eid.net.ClaimService;
import ee.cyber.eid.util.Util;
import ee.sk.digidoc.DigiDocException;
import ee.sk.utils.ConfigManager;
import ee.sk.xmlenc.EncryptedData;

/**
 * A desktop client for communicating with the claim handling service.
 * Usage: java EidClient [url] [-cfg config]
 */
public class EidClient {

    /** Default place to look for the configuration file. */
    private static final String DEFAULT_CONF = "jdigidoc.cfg";

    /** The default location of our claim handling service. */
    private static final String DEFAULT_URL = "https://localhost/eid/avaldused/submit.php";

    /** The default token index to use, i.e. 0 means we use the first card
     * reader we find. */
    private static final int TOKEN_INDEX = 0;

    public static void main(String[] argv) throws Exception {
        String cfg = DEFAULT_CONF;
        String url = DEFAULT_URL;

        /* Parse the arguments. */
        for (int i = 0; i < argv.length; i++) {
            if ("-cfg".equals(argv[i])) {
                if (++i < argv.length) {
                    cfg = argv[i];
                } else {
                    System.err.println("No parameter given after '-cfg'.");
                    return;
                }
            } else {
                url = argv[i];
            }
        }

        /* Check the configuration. */
        if (!ConfigManager.init(cfg)) {
            System.err.println("Add " + DEFAULT_CONF + " to the current "
                    + "directory or specify a configuration file as a "
                    + "command-line argument.");
            return;
        }

        /* Check if we have card terminals. */
        if (CardUtil.getTerminals().isEmpty()) {
            throw new CardException("No card terminals found");
        }

        /* Identify the user. */
        System.out.println("Identifying the user...");
        X509Certificate cert = Authenticator.authenticate(TOKEN_INDEX);
        System.out.println("Ok.");

        /* Find the subject's name.
         * We can't use ConvertUtils.getCommonName() from jdigidoc, because it
         * only returns the CN until the first ',' - on personal cards this
         * gives us only the surname. */
        String cn = Util.getSubjectCN(cert);
        String cnTokens[] = cn.split(",");
        String subject = (cnTokens.length != 3) ? cn
                : cnTokens[1] + ' ' + cnTokens[0] + ", " + cnTokens[2];

        /* Let the user insert something to submit. */
        System.out.print("Enter claim content: ");
        Scanner scanner = new Scanner(System.in);
        String content = scanner.nextLine();

        /* Create the claim, sign and verify it. */
        System.out.println("Reading personal data...");
        Claim claim = createClaim(subject + '\n' + content, TOKEN_INDEX);
        System.out.println("Signing the claim...");
        claim.sign(TOKEN_INDEX);
        if (claim.verify()) {
            System.out.println("Ok.");
        } else {
            System.err.println("Signing failed.");
            return;
        }

        /* Save the claim to disc for reference. */
        claim.saveTo("claim.ddoc");

        /* Submit the claim and get a response. */
        System.out.println("Submitting the claim...");
        EncryptedData encd = ClaimService.submit(url, claim, cert);
        if (validateEncryptedData(encd)) {
            System.out.println("Ok, got a valid response.");
        } else {
            System.err.println("Got an invalid response.");
            return;
        }

        /* Parse and verify the response. */
        System.out.println("Decrypting and processing the response...");
        Response resp = new Response(encd);
        resp.decrypt(cn, TOKEN_INDEX);
        try {
            resp.verify();
        } catch (Exception e) {
            System.err.println("Response signature verification failed: " + e.toString());
        }
        String responseContent = resp.getContent();
        if (responseContent == null) {
            System.err.println("The response is empty!");
            return;
        }
        System.out.println("Ok.");

        System.out.println("The service responded:");
        System.out.println(responseContent);
    }

    /** Create a new claim. */
    private static Claim createClaim(String content, int terminalIndex)
            throws Exception {
        /* Add the given content to a container. */
        Claim claim = new Claim();
        claim.addClaimFile(content);

        /* Read personal data from the card. */
        EstEID eid = new EstEID(terminalIndex);
        claim.addPersonalDataFile(eid.getPersonalData());

        return claim;
    }

    /** Validate an EncryptedData object. */
    @SuppressWarnings("unchecked")
    private static boolean validateEncryptedData(EncryptedData encd) {
        List<DigiDocException> errs = encd.validate();
        if (errs.size() > 0) {
            System.err.println("Could not validate response:");
            for (DigiDocException e : errs) {
                System.err.println(e.getMessage());
            }
            return false;
        }
        return true;
    }

}

File: avaldused/client-dotnet/EidClient.cs
using System;
using System.IO;
using System.Security.Cryptography.X509Certificates;
using System.Text.RegularExpressions;

namespace EidClient
{
    /// <summary>
    /// Command line tool that authenticates the user, creates a claim and
    /// sends it to the service. The response's signature is checked and
    /// the response text is presented to the user.
    /// </summary>
    class EidClient
    {
        /// <summary>
        /// Default location of the clam handling service.
        /// </summary>
        private static string DEFAULT_URL = "https://localhost/eid/submit.php";

        [STAThread]
        static void Main(string[] args)
        {
            /* Take the location of the claim handling service as an optional parameter. */
            String url = DEFAULT_URL;
            if (args.Length > 0)
                url = args[0];

            Console.Write("Authenticating user...");
            X509Certificate2 cert = Authenticator.authenticate();
            Console.WriteLine("Ok.");

            Console.Write("Enter claim content: ");
            string content = Console.ReadLine();
            Claim claim = new Claim();
            claim.addContent(getSubject(cert) + '\n' + content);

            Console.Write("Reading personal data...");
            claim.addContent(PersonalData.ToString());
            Console.WriteLine("Ok.");

            Console.Write("Signing the claim...");
            claim.sign();
            Console.WriteLine("Ok.");

            Console.Write("Saving the document to disk...");
            string claimFile = Path.GetTempFileName();
            claim.saveToFile(claimFile);
            Console.WriteLine("Ok.");

            Console.Write("Submitting the claim...");
            byte[] responseBytes = ClaimService.submit(url, claimFile);
            Console.Write("Ok.");

            Response resp = new Response(responseBytes);
            resp.verify();
            Console.WriteLine("The service answered: " + resp.getContent());

            /* Pause the execution, so people have a chance to read the output. */
            Console.Write("Press Return to exit...");
            Console.ReadLine();
            return;
        }

        /// <summary>
        /// Returns the subject's name from the certificate in the form "GivenName Surname, Serial".
        /// </summary>
        /// <param name="cert">The subject's certificate.</param>
        /// <returns>The subject's formatted CN or the complete DN if pattern matching failed.</returns>
        private static string getSubject(X509Certificate2 cert)
        {
            /* Search for the Common Name and capture parts of it into groups. Alternatively you could
             * search for the values of SN, G and SERIALNUMBER.
             * For each match, group 0 will be the whole CN, group 1 the surname, group 2 the given name
             * and group 3 the serial number. */
            string pattern = "CN=\"(.+),(.+),(\\d+)\"";
            string DN = cert.Subject;
            foreach (Match match in Regex.Matches(DN, pattern))
                /* Return the first match. */
                return match.Groups[2].Value + ' ' + match.Groups[1].Value + ", " + match.Groups[3].Value;

            /* If no match found, return the whole DN. */
            return DN;
        }
    }
}
